zend_abstract_interface/functions/php8/functions.c

```diff
@@ -0,0 +1,139 @@
+
+    zend_function *func = NULL;
+    zend_string *zsname = NULL;
+    ALLOCA_FLAG(use_heap)
+
+    /* We look up the function handler directly from zend_fetch_function()
+     * instead of calling zend_is_callable_ex() because:
+     *
+     *   - We always call functions by their string name and not one of the
+     *     many 'callable' types, so the code path can be simplified.
+     *   - We assume the caller always provides the lowercased function name so
+     *     we can avoid zend_str_tolower_copy().
+     *   - We assume the caller will not prefix the function name with a
+     *     root-scope prefix, '\'.
+     *
+     * In addition, we look up the function handler from zend_fetch_function()
+     * instead of directly from the EG(function_table) to ensure that the
+     * runtime cache for userland functions will be initialized.
+     */
+    ZSTR_ALLOCA_ALLOC(zsname, name_len, use_heap);
+    memcpy(ZSTR_VAL(zsname), name, (name_len + 1));
+    func = zend_fetch_function(zsname);
+    ZSTR_ALLOCA_FREE(zsname, use_heap);
+
+    /* "function %s() does not exist" */
+    if (!func) return false;
+
+    zend_fcall_info_cache fcc = {0};
+    fcc.function_handler = func;
+
+    zend_fcall_info fci = {0};
+    fci.size = sizeof(zend_fcall_info);
+    fci.retval = retval;
+
+    /* We add the zval args directly from the variable arguments API instead of
+     * using zend_fcall_info_argv() which allows us to NULL-check each arg. It
+     * also allows us to avoid an unnecessary call to
+     * zend_fcall_info_args_clear() since the 'param{s|_count}' members in
+     * 'zend_fcall_info' will always be zeroed out in our case.
+     *
+     * Based on zend_fcall_info_argv():
+     * https://github.com/php/php-src/blob/PHP-8.0.4/Zend/zend_API.c?#L3608-L3622
+     */
+    if (argc > 0) {
+        va_list argv;
+        va_start(argv, argc);
+
+        zval *arg;
+        fci.param_count = (uint32_t)argc;
+        fci.params = (zval *)ecalloc(1, (fci.param_count * sizeof(zval)));
+
+        for (uint32_t i = 0; i < (uint32_t)argc; ++i) {
+            arg = va_arg(argv, zval *);
+            if (!arg) {
+                /* Only dtor the zvals that were successfully copied */
+                fci.param_count = i;
+                zend_fcall_info_args_clear(&fci, /* free_mem */ true);
+                return false;
+            }
+            ZVAL_COPY(&fci.params[i], arg);
+        }
+
+        va_end(argv);
+    }
+
+    bool ret = false;
+
+    zai_sandbox sandbox;
+    zai_sandbox_open(&sandbox);
+
+    zend_try {
+        zend_result result = zend_call_function(&fci, &fcc);
+        /* An unhandled exception will not result in a zend_bailout if there is
+         * an active execution context. This is a failed call if an exception
+         * was thrown. The sandbox will clean up our mess when it closes.
+         */
+        ret = (result == SUCCESS && !EG(exception));
```

Comment:
Not a change required for this PR, but rather an observation for the design of future zai modules. Do you think that a binary return state is enough to communicate the outcome? Or it would make sense to have at least 3 states:

1. `SUCCESS`;
2. error because of something you did wrong while developing (e.g. calling with a `NULL` return value);
3. runtime error (e.g. and exception because the database is not reachable).

Reply:
> Do you think that a binary return state is enough to communicate the outcome?

It is not.

> Or it would make sense to have at least 3 states:

Very good point. [ZAI error types are on the wishlist](https://github.com/DataDog/dd-trace-php/blob/f417c10/zend_abstract_interface/sandbox/sandbox.h#L7-L12). ([APMPHP-614](https://datadoghq.atlassian.net/browse/APMPHP-614)) They will likely be tightly coupled to the sandbox so that exception and error states can be moved out of the PHP global space and into a locally-managed error state.

The return states would be similar to what you proposed. Off the top of my head, this will include the following basic return states:

0. **Success**
1. **Bailout**; Allows the caller to clean up before bubbling up a `zend_bailout`.
2. **Failure**; This can have a number of generic sub types (e.g. exception thrown, error raised, NULL argument, etc) and API-specific types (e.g. method does not exist).

Reply:
Thanks @SammyK for clarifying